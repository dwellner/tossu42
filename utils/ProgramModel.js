import DateUtils from "./DateUtils";
import ProgramService from "../data/ProgramService";

const getIVTotalDistance = (ivDistance, repeats) => {
  const warmup = 2,
    coolDown = 1,
    recovery = 0.4;
  return warmup + (ivDistance + recovery) * repeats + coolDown;
};

const getDayDistance = day => {
  if (day.type === "iv") return getIVTotalDistance(day.distance, day.repeat);
  if (day.type === "lepo") return 0;
  return day.distance;
};

const getWeekDistance = week =>
  week.map(getDayDistance).reduce((a, b) => a + b);

const getStretchedProgramStart = (program, programLength, totalLength) => {
  let repeats = [];

  const stretchLengthInWeeks = Math.ceil(totalLength - programLength);
  const repeatBlock = program.weeks.slice(0, program.stretchRules.repeat);
  const numberOfRepeatBlocks = Math.ceil(
    stretchLengthInWeeks / repeatBlock.length
  );
  for (i = 0; i < numberOfRepeatBlocks; i++) {
    repeats = repeats.concat(repeatBlock.map(days => [].concat(days)));
  }
  const excessWeeks = repeats.length - stretchLengthInWeeks;
  return repeats.slice(excessWeeks);
};

const getProgramWeeks = (program, length) => {
  const programLength = program.weeks.length;

  if (length === programLength || program.stretchRules === undefined) {
    return program.weeks;
  }

  if (length > programLength) {
    const stretchedStart = getStretchedProgramStart(
      program,
      programLength,
      length
    );
    return [].concat(stretchedStart, program.weeks);
  }

  // TODO: check total length is within program range, or return mimimum
  return []; // TODO: handle shrinking
};

export default {
  getWeekProgram: (targetDate, { targetTime, name, length }) => {
    if (targetDate === null) return null;

    const programs = ProgramService.getPrograms(targetTime, name, length);
    if (programs.length === 0) return null;
    const program = programs[0];

    const programWeeks = getProgramWeeks(program, length);
    const firstDate = DateUtils.nextDate(
      targetDate,
      programWeeks.length * -7 + 1
    );

    let i = 0;
    const weeks = programWeeks.map((week, index) => {
      const weekNumber = programWeeks.length - index;
      const distance = getWeekDistance(week);
      const days = week.map(day => ({
        ...day,
        date: DateUtils.nextDate(firstDate, i++)
      }));
      return { weekNumber, distance, days };
    });
    return { weeks };
  }
};
