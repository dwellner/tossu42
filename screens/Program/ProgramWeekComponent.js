import React from "react";
import PropTypes from "prop-types";

import { Text, View, StyleSheet } from "react-native";
import Styles from "../../constants/Styles";
import Colors from "../../constants/Colors";
import { dateToDate, dateRangetoLabel } from "../../utils/Formatters";
import { Labels } from "../../constants/Texts";

const RaceInfo = ({ targetEvent }) => (
  <View style={styles.raceInfo}>
    <Text style={{ ...Styles.strongContent }}>{targetEvent.name}</Text>
    <Text style={{ ...Styles.defaultContent }}>
      {dateToDate(targetEvent.date)}
    </Text>
  </View>
);

const LabeledValue = ({ label, value, valueStyle, isPast }) => {
  const color = isPast ? Colors.inactiveText : Colors.defaultText;
  return (
    <View style={{ margin: 16 }}>
      <Text style={{ ...Styles.lightContent, fontSize: 12, color }}>
        {label}
      </Text>
      <Text style={{ ...valueStyle, color }}>{value}</Text>
    </View>
  );
};

export default class ProgramWeekComponent extends React.PureComponent {
  static propTypes = {
    weekNumber: PropTypes.number.isRequired,
    targetEvent: PropTypes.object.isRequired,
    week: PropTypes.object.isRequired,
    isPast: PropTypes.bool.isRequired
  };

  render() {
    const { weekNumber, targetEvent, week, isPast } = this.props;
    const race =
      weekNumber == 1 ? <RaceInfo targetEvent={targetEvent} /> : undefined;
    const weekDates = dateRangetoLabel([week.days[0].date, week.days[6].date]);
    const distance = week.distance.toFixed(0);

    return (
      <View>
        <View style={styles.weekInfoRow}>
          <LabeledValue
            isPast={isPast}
            label={Labels.weekShort}
            value={weekNumber}
            valueStyle={Styles.strongLargeContent}
          />
          <LabeledValue
            isPast={isPast}
            label={Labels.eventDate}
            value={weekDates}
            valueStyle={Styles.largeContent}
          />
          <LabeledValue
            isPast={isPast}
            label={"km"}
            value={distance}
            valueStyle={Styles.strongLargeContent}
          />
        </View>
        {race}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  raceInfo: {
    margin: 16,
    flex: 1,
    alignItems: "center"
  },

  weekInfoRow: {
    flex: 1,
    flexDirection: "row",
    justifyContent: "space-between"
  }
});
