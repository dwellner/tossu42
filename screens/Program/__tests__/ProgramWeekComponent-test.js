import "react-native";
import React from "react";
import ProgramWeekComponent from "../ProgramWeekComponent";
import renderer from "react-test-renderer";
import ProgramModel from "../../../utils/ProgramModel";

it("renders correctly", () => {
  const targetEvent = { name: "Foo maraton", date: "2018-10-30" };

  const weekProgram = ProgramModel.getWeekProgram("2018-10-30", {
    targetTime: 240,
    name: "Peruskunto - 4t tavoiteaika",
    length: 24
  });
  const tree = renderer
    .create(
      <ProgramWeekComponent
        weekNumber={4}
        week={weekProgram.weeks[3]}
        isPast={false}
        targetEvent={targetEvent}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
