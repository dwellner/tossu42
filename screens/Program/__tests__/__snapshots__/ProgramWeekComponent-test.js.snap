// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "margin": 16,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#535353",
            "fontFamily": "OpenSans-Light",
            "fontSize": 12,
          }
        }
      >
        Viikko
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#535353",
            "fontFamily": "OpenSans-Bold",
            "fontSize": 20,
          }
        }
      >
        4
      </Text>
    </View>
    <View
      style={
        Object {
          "margin": 16,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#535353",
            "fontFamily": "OpenSans-Light",
            "fontSize": 12,
          }
        }
      >
        Pvm
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#535353",
            "fontFamily": "OpenSans",
            "fontSize": 20,
          }
        }
      >
        6.6.18 - 12.6.18
      </Text>
    </View>
    <View
      style={
        Object {
          "margin": 16,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#535353",
            "fontFamily": "OpenSans-Light",
            "fontSize": 12,
          }
        }
      >
        km
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#535353",
            "fontFamily": "OpenSans-Bold",
            "fontSize": 20,
          }
        }
      >
        46
      </Text>
    </View>
  </View>
</View>
`;
